File: platforms/documentation/docs/src/samples/templates/gradle-plugin-in-java/src/main/java/com/example/plugin/GreetingPlugin.java
package com.example.plugin;

import org.gradle.api.Plugin;
import org.gradle.api.Project;

public class GreetingPlugin implements Plugin<Project> {
    public void apply(Project project) {
        project.getTasks().register("greet", task -> {
            task.doLast(s -> System.out.println("Hello from plugin 'com.example.plugin.greeting'"));
        });
    }
}

File: platforms/documentation/docs-site/src/samples/reference/testing-gradle-plugins/plugin-binary/kotlin/plugin/src/main/kotlin/org/example/GreetingPlugin.kt
package org.example

import org.gradle.api.Plugin
import org.gradle.api.Project

class GreetingPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        // Create extension and set sensible defaults (conventions)
        val ext = project.extensions.create("greeting", GreetingExtension::class.java).apply {  // <1>
            message.convention("Hello from plugin")
            outputFile.convention(project.layout.buildDirectory.file("greeting.txt"))
        }

        // Register task and map extension -> task inputs (lazy wiring)
        project.tasks.register("greet", GreetTask::class.java) {    // <2>
            group = "example"
            description = "Writes a greeting to a file"
            message.set(ext.message)
            outputFile.set(ext.outputFile)
        }
    }
}
